### PR TITLE
[REST API] Product Reports

### LINKED ISSUE #20775: REST API: Products reports endpoint
Schema:

- https://app.swaggerhub.com/apis/peterfabian/wc-v3-api/1.0.0#/default/get_reports_products
- https://app.swaggerhub.com/apis/peterfabian/wc-v3-api/1.0.0#/default/get_reports_products_stats

### PR DESCRIPTION
**Pull request by claudiulodro**
Closes #20775.

### CODE REVIEW

**Comment by claudiulodro** [comment]
To finish up the `reports/products` endpoint in this PR requires #20936

File: includes/api/class-wc-rest-reports-products-controller.php [hunk 1]
```diff
@@ -31,4 +31,194 @@ class WC_REST_Reports_Products_Controller extends WC_REST_Reports_Controller {
...
+					'context'     => array( 'view', 'edit' ),
+					'description' => __( 'Total gross revenue of all items sold.', 'woocommerce' ),
+				),
+				'orders_count' => array(
+					'type'        => 'integer',
+					'readonly'    => true,
+					'context'     => array( 'view', 'edit' ),
+					'description' => __( 'Number of orders product appeared in.', 'woocommerce' ),
+				),
+				'_links' => array(
```
Comment (claudiosanches) [inline includes/api/class-wc-rest-reports-products-controller.php]: We don't include `_links` inside the schema, WP core does not include.

**Comment by claudiulodro** [comment]
@peterfabian I think something similar to https://github.com/woocommerce/woocommerce/pull/20956/commits/ae631108069744898178b024d26869c088525b4b is going to be the way to go for the `products/stats` endpoint: A new data store that aggregates info from the `wc_order_product_lookup` table in a different way than `WC_Reports_Products_Data_Store`.

We'll probably want to move all the new reports data stores into their own subfolder when we're through with the stats API. :)

File: includes/api/class-wc-rest-reports-products-controller.php [hunk 2]
```diff
@@ -31,4 +31,199 @@ class WC_REST_Reports_Products_Controller extends WC_REST_Reports_Controller {
...
+	}
+
+	/**
+	 * Prepare links for the request.
+	 *
+	 * @param WC_Data         $object  Object data.
+	 * @param WP_REST_Request $request Request object.
+	 * @return array                   Links for the given post.
+	 */
+	protected function prepare_links( $object, $request ) {
```
Comment (claudiosanches) [inline includes/api/class-wc-rest-reports-products-controller.php]: `$request` is not used here, we include in some endpoints when required to fetch IDs or something else, but here we can remove it.

Comment (peterfabian) [reply includes/api/class-wc-rest-reports-products-controller.php]: Thanks, fixed.

**Comment by claudiulodro** [comment]
This is working well so far. Just 2 last fixes then this can be merged:
- `Warning: Declaration of WC_Reports_Products_Stats_Data_Store::get_sql_query_params($query_args, &$totals_params, &$intervals_params) should be compatible with WC_Reports_Products_Data_Store::get_sql_query_params($query_args) in /srv/www/woodev/wp-content/plugins/woocommerce/includes/data-stores/class-wc-reports-products-stats-data-store.php on line 17`
- `reports/products/stats` API response `subtotals` section is an stdClass object instead of an array

To reproduce: `phpunit tests/unit-tests/api/reports-products-stats`

I'll fix them tomorrow when I have a chance if these bugs are still around so we can get this merged, otherwise anyone can. :)

**Comment by peterfabian** [comment]
Created a branch off this so that we can decide if we want to include it or not: #21045.

**Comment by peterfabian** [comment]
I will merge this so that we can continue working on other endpoints, there have been quite a few updates to the base classes which we build on top of.

**Pull request merged**